pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,336 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+// Package scheduler features the scheduler for Fleet workloads.
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	fleetv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/scheduler/framework"
+	"go.goms.io/fleet/pkg/scheduler/queue"
+	"go.goms.io/fleet/pkg/utils/controller"
+)
+
+// Scheduler is the scheduler for Fleet workloads.
+type Scheduler struct {
+	// name is the name of the scheduler.
+	name string
+
+	// framework is the scheduling framework in use by the scheduler.
+	//
+	// At this stage, a scheduler is always associated with one scheduling framework; in the long
+	// run, multiple frameworks may be supported, to allow the usage of varying scheduling configurations
+	// for different types of workloads.
+	framework framework.Framework
+
+	// queue is the work queue in use by the scheduler; the scheduler pulls items from the queue and
+	// performs scheduling in accordance with them.
+	queue queue.ClusterResourcePlacementSchedulingQueue
+
+	// client is the (cached) client in use by the scheduler for accessing Kubernetes API server.
+	client client.Client
+	// uncachedReader is the uncached read-only client in use by the scheduler for accessing
+	// Kubernetes API server; in most cases client should be used instead, unless consistency becomes
+	// a serious concern.
+	//
+	// TO-DO (chenyu1): explore the possbilities of using a mutation cache for better performance.
+	uncachedReader client.Reader
+	// manager is the controller manager in use by the scheduler.
+	manager ctrl.Manager
+	// eventRecorder is the event recorder in use by the scheduler.
+	eventRecorder record.EventRecorder
+}
+
+// NewScheduler creates a scheduler.
+func NewScheduler(
+	name string,
+	framework framework.Framework,
+	queue queue.ClusterResourcePlacementSchedulingQueue,
+	manager ctrl.Manager,
+) *Scheduler {
+	return &Scheduler{
+		name:           name,
+		framework:      framework,
+		queue:          queue,
+		client:         manager.GetClient(),
+		uncachedReader: manager.GetAPIReader(),
+		manager:        manager,
+		eventRecorder:  manager.GetEventRecorderFor(name),
+	}
+}
+
+// ScheduleOnce performs scheduling for one single item pulled from the work queue.
+func (s *Scheduler) scheduleOnce(ctx context.Context) {
```

Comment:
please add a TODO to add  scheduling metrics

Reply:
Added.